### PR TITLE
[WIP] Compile runtime with -fvisibility=hidden

### LINKED ISSUE #2484: Compile runtime with -fvisibility=hidden
-fvisibility=hidden enables several important optimizations:
- Access to static variables will be faster, without GOT indirection. We can get rid of PREPARE_EXTERNAL_VAR on x64 and instead access the static variables directly, like on Windows.
- Access to current thread object will be faster. it is known bottleneck in allocation helpers.

### PR DESCRIPTION
**Pull request by fiigii**
Based on @janvorli 's work to resolve #2484.

### CODE REVIEW

File: src/nativeresources/resourcestring.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ int LoadNativeStringResource(const NativeStringResourceTable &nativeStringResour
 #define NATIVE_STRING_RESOURCE_TABLE(name) CONCAT(nativeStringResourceTable_, name)
 
 #define DECLARE_NATIVE_STRING_RESOURCE_TABLE(name) \
-    extern const NativeStringResourceTable NATIVE_STRING_RESOURCE_TABLE(name)
+    __attribute__((visibility("default"))) extern const NativeStringResourceTable NATIVE_STRING_RESOURCE_TABLE(name)
```
Comment (fiigii) [inline src/nativeresources/resourcestring.h]: Still has the error message:
```
Undefined symbols for architecture x86_64:
  "_nativeStringResourceTable_mscorrc_debug", referenced from:
```
@jkotas @janvorli Any suggestion?

Comment (fiigii) [inline src/nativeresources/resourcestring.h]: Still has the error message:
```
Undefined symbols for architecture x86_64:
  "_nativeStringResourceTable_mscorrc_debug", referenced from:
```
@jkotas @janvorli @AndyAyersMS  Any suggestion?

Comment (jkotas) [reply src/nativeresources/resourcestring.h]: @janvorli has the best context to comment on what may be going on here. He is on the plane right now. He should be able to respond tomorrow.

Comment (fiigii) [reply src/nativeresources/resourcestring.h]: @jkotas Thank you!

Comment (janvorli) [reply src/nativeresources/resourcestring.h]: @fiigii it looks like the attribute should be placed at the end of the declaration for variables. So
```C++
extern const NativeStringResourceTable NATIVE_STRING_RESOURCE_TABLE(name) __attribute__((visibility("default")))
```
I didn't have a chance to try it yet though.

Comment (fiigii) [reply src/nativeresources/resourcestring.h]: @janvorli Yes, you are right https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Variable-Attributes.html 
Thank you so much!

**Comment by fiigii** [comment]
Addressed all the errors and warnings from compile and linke stage.

Now, checked tests show the assert failure when run tests:
```
Assert failure(PID 51348 [0x0000c894], Thread: 46139232 [0x2c00760]): !"Recursion in CLRException::GetThrowable"
File: /Users/dotnet-bot/j/workspace/dotnet_coreclr/master/checked_osx10.12_prtest/src/vm/clrex.cpp Line: 149
Image: /Users/dotnet-bot/j/workspace/dotnet_coreclr/master/checked_osx10.12_tst_prtest/bin/tests/Windows_NT.x64.Checked/Tests/coreoverlay/corerun
```
OSX provides the stack frame information:
```
* thread #1: tid = 0x0000, 0x00007fffa24b6dd6 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fffa24b6dd6 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fffa25a2787 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fffa241c420 libsystem_c.dylib`abort + 129
    frame #3: 0x000000010ba6923e libcoreclr.dylib`PROCAbort + 14
    frame #4: 0x000000010ba65f3a libcoreclr.dylib`PROCEndProcess(void*, unsigned int, int) + 762
    frame #5: 0x000000010ba66008 libcoreclr.dylib`TerminateProcess + 152
    frame #6: 0x000000010bd733a7 libcoreclr.dylib`SafeExitProcess(unsigned int, int, ShutdownCompleteAction) + 551
    frame #7: 0x000000010bd74877 libcoreclr.dylib`EEPolicy::HandleFatalError(unsigned int, unsigned long, char16_t const*, _EXCEPTION_POINTERS*) + 375
    frame #8: 0x000000010bff5abd libcoreclr.dylib`SystemNative::GenericFailFast(REF<StringObject>, REF<ExceptionObject>, unsigned long, unsigned int) + 445
    frame #9: 0x000000010bff5d44 libcoreclr.dylib`SystemNative::FailFast(StringObject*) + 580
    frame #10: 0x00000001129ae39c
    frame #11: 0x00000001129ae185
    frame #12: 0x0000000112b89ded
    frame #13: 0x000000010bf50b47 libcoreclr.dylib`CallDescrWorkerInternal + 124
    frame #14: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #15: 0x000000010bd23cca libcoreclr.dylib`DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 106
    frame #16: 0x000000010bd24d85 libcoreclr.dylib`CallDefaultConstructor(OBJECTREF) + 517
    frame #17: 0x000000010bd2d7bd libcoreclr.dylib`EEException::CreateThrowable() + 269
    frame #18: 0x000000010bbf0a30 libcoreclr.dylib`CreateCOMPlusExceptionObject(Thread*, _EXCEPTION_RECORD*, int) + 880
    frame #19: 0x000000010beb71ea libcoreclr.dylib`ExceptionTracker::CreateThrowable(_EXCEPTION_RECORD*, int) + 106
    frame #20: 0x000000010beb1979 libcoreclr.dylib`ExceptionTracker::GetOrCreateTracker(unsigned long, StackFrame, _EXCEPTION_RECORD*, _CONTEXT*, int, bool, ExceptionTracker::StackTraceState*) + 1545
    frame #21: 0x000000010beb0a2a libcoreclr.dylib`ProcessCLRException + 1018
    frame #22: 0x000000010beb7e09 libcoreclr.dylib`UnwindManagedExceptionPass1(PAL_SEHException&, _CONTEXT*) + 537
    frame #23: 0x000000010beb82e6 libcoreclr.dylib`DispatchManagedException(PAL_SEHException&, bool) + 198
    frame #24: 0x000000010beaee3a libcoreclr.dylib`HandleHardwareException(PAL_SEHException*) + 730
    frame #25: 0x000000010b9df79a libcoreclr.dylib`SEHProcessException(PAL_SEHException*) + 346
    frame #26: 0x000000010ba7348b libcoreclr.dylib`PAL_DispatchException + 283
    frame #27: 0x000000010ba72b85 libcoreclr.dylib`PAL_DispatchExceptionWrapper + 10
    frame #28: 0x00000001129c59e7
    frame #29: 0x0000000112a42978
    frame #30: 0x0000000112a3fd2a
    frame #31: 0x0000000112a3fa13
    frame #32: 0x00000001129f6b2e
    frame #33: 0x00000001129f83aa
    frame #34: 0x00000001129f8774
    frame #35: 0x0000000112a10c03
    frame #36: 0x00000001129cd3b3
    frame #37: 0x00000001129cd085
    frame #38: 0x00000001129ae513
    frame #39: 0x00000001129ae185
    frame #40: 0x0000000112b89ded
    frame #41: 0x000000010bf50b47 libcoreclr.dylib`CallDescrWorkerInternal + 124
    frame #42: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #43: 0x000000010bd23cca libcoreclr.dylib`DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 106
    frame #44: 0x000000010bd24d85 libcoreclr.dylib`CallDefaultConstructor(OBJECTREF) + 517
    frame #45: 0x000000010bd2d7bd libcoreclr.dylib`EEException::CreateThrowable() + 269
    frame #46: 0x000000010bbf0a30 libcoreclr.dylib`CreateCOMPlusExceptionObject(Thread*, _EXCEPTION_RECORD*, int) + 880
    frame #47: 0x000000010beb71ea libcoreclr.dylib`ExceptionTracker::CreateThrowable(_EXCEPTION_RECORD*, int) + 106
    frame #48: 0x000000010beb1979 libcoreclr.dylib`ExceptionTracker::GetOrCreateTracker(unsigned long, StackFrame, _EXCEPTION_RECORD*, _CONTEXT*, int, bool, ExceptionTracker::StackTraceState*) + 1545
    frame #49: 0x000000010beb0a2a libcoreclr.dylib`ProcessCLRException + 1018
    frame #50: 0x000000010beb7e09 libcoreclr.dylib`UnwindManagedExceptionPass1(PAL_SEHException&, _CONTEXT*) + 537
    frame #51: 0x000000010beb82e6 libcoreclr.dylib`DispatchManagedException(PAL_SEHException&, bool) + 198
    frame #52: 0x000000010beaee3a libcoreclr.dylib`HandleHardwareException(PAL_SEHException*) + 730
    frame #53: 0x000000010b9df79a libcoreclr.dylib`SEHProcessException(PAL_SEHException*) + 346
    frame #54: 0x000000010ba7348b libcoreclr.dylib`PAL_DispatchException + 283
    frame #55: 0x000000010ba72b85 libcoreclr.dylib`PAL_DispatchExceptionWrapper + 10
    frame #56: 0x00000001129c59e7
    frame #57: 0x0000000112a42978
    frame #58: 0x0000000112a3fd2a
    frame #59: 0x0000000112a3fa13
    frame #60: 0x00000001129f6b2e
    frame #61: 0x00000001129f83aa
    frame #62: 0x00000001129f8774
    frame #63: 0x0000000112a10c03
    frame #64: 0x00000001129cd3b3
    frame #65: 0x00000001129cd085
    frame #66: 0x00000001129ae513
    frame #67: 0x00000001129ae185
    frame #68: 0x0000000112b852bd
    frame #69: 0x000000010bf50b47 libcoreclr.dylib`CallDescrWorkerInternal + 124
    frame #70: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #71: 0x000000010bd23cca libcoreclr.dylib`DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 106
    frame #72: 0x000000010bd24d85 libcoreclr.dylib`CallDefaultConstructor(OBJECTREF) + 517
    frame #73: 0x000000010bd2d7bd libcoreclr.dylib`EEException::CreateThrowable() + 269
    frame #74: 0x000000010bd2a212 libcoreclr.dylib`CLRException::GetThrowable() + 994
    frame #75: 0x000000010bd2b335 libcoreclr.dylib`CLRException::GetThrowableFromException(Exception*) + 133
    frame #76: 0x000000010bbea0e0 libcoreclr.dylib`UnwindAndContinueRethrowHelperInsideCatch(Frame*, Exception*) + 96
    frame #77: 0x000000010bbcc3dd libcoreclr.dylib`NDirectImportWorker + 861
    frame #78: 0x000000010bf51af4 libcoreclr.dylib`NDirectImportThunk + 70
    frame #79: 0x0000000112bbd362
    frame #80: 0x0000000112bbd3de
    frame #81: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #82: 0x000000010bd23cca libcoreclr.dylib`DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 106
    frame #83: 0x000000010bd23e43 libcoreclr.dylib`DispatchCallSimple(unsigned long*, unsigned int, unsigned long, unsigned int) + 163
    frame #84: 0x000000010bc613de libcoreclr.dylib`MethodTable::RunClassInitEx(OBJECTREF*) + 590
    frame #85: 0x000000010bc6203e libcoreclr.dylib`MethodTable::DoRunClassInitThrowing() + 2126
    frame #86: 0x000000010bc631aa libcoreclr.dylib`MethodTable::CheckRunClassInitThrowing() + 746
    frame #87: 0x000000010bdc4ee8 libcoreclr.dylib`JIT_GetSharedNonGCStaticBase_Helper + 552
    frame #88: 0x0000000112a3f84a
    frame #89: 0x0000000112a3f8c6
    frame #90: 0x00000001129f6b2e
    frame #91: 0x00000001129f688d
    frame #92: 0x00000001129f8bac
    frame #93: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #94: 0x000000010bd23cca libcoreclr.dylib`DispatchCallDebuggerWrapper(CallDescrData*, ContextTransitionFrame*, int) + 106
    frame #95: 0x000000010bd23e43 libcoreclr.dylib`DispatchCallSimple(unsigned long*, unsigned int, unsigned long, unsigned int) + 163
    frame #96: 0x000000010bc613de libcoreclr.dylib`MethodTable::RunClassInitEx(OBJECTREF*) + 590
    frame #97: 0x000000010bc6203e libcoreclr.dylib`MethodTable::DoRunClassInitThrowing() + 2126
    frame #98: 0x000000010bc631aa libcoreclr.dylib`MethodTable::CheckRunClassInitThrowing() + 746
    frame #99: 0x000000010bdc54d8 libcoreclr.dylib`JIT_GetSharedGCStaticBase_Helper + 552
    frame #100: 0x00000001129f8959
    frame #101: 0x00000001129e54dd
    frame #102: 0x0000000113030a0b
    frame #103: 0x000000010bf50b47 libcoreclr.dylib`CallDescrWorkerInternal + 124
    frame #104: 0x000000010bd23c2f libcoreclr.dylib`CallDescrWorkerWithHandler(CallDescrData*, int) + 527
    frame #105: 0x000000010bd24b13 libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) + 2003
    frame #106: 0x000000010bb49548 libcoreclr.dylib`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) + 1192
    frame #107: 0x000000010bb49900 libcoreclr.dylib`Assembly::ExecuteMainMethod(REF<PtrArray>*, int) + 528
    frame #108: 0x000000010bbac26f libcoreclr.dylib`CorHost2::ExecuteAssembly(unsigned int, char16_t const*, int, char16_t const**, unsigned int*) + 751
    frame #109: 0x000000010ba79d7d libcoreclr.dylib`coreclr_execute_assembly + 381
    frame #110: 0x000000010b98de38 corerun`ExecuteManagedAssembly(char const*, char const*, char const*, int, char const**) + 1176
    frame #111: 0x000000010b98d1ff corerun`corerun(int, char const**) + 335
    frame #112: 0x00007fffa2388255 libdyld.dylib`start + 1
    frame #113: 0x00007fffa2388255 libdyld.dylib`start + 1
```

**Comment by fiigii** [comment]
@janvorli @jkotas @AndyAyersMS Any suggestion to solve it?

**Comment by jkotas** [comment]
The problem is likely in NDirectImportWorker. Take a look at the name of the method it is trying to load.

**Comment by janvorli** [comment]
@fiigii I was wondering if you are planning to finish this.

**Comment by fiigii** [comment]
@janvorli Sorry, I am busy on the hardware intrinsic work recently, I maybe do not have time to finish this before 2.1.

**Comment by fiigii** [comment]
@janvorli Does this have to be shiped in .NET Core 2.1?

**Comment by janvorli** [comment]
@fiigii I don't think it has to. The hardware intrinsics seem more important. I was mostly wondering if you plan to continue working on it later or not.

**Comment by fiigii** [comment]
> I was mostly wondering if you plan to continue working on it later or not.

Yes, I will  continue working on it later if this is not very urgent.

**Comment by kasper3** [comment]
@fiigii, are you planning to continue working on it for next version?

**Comment by fiigii** [comment]
@kasper3 Yes, I will, but maybe after finishing the hardware intrinsic project.

**Comment by jkotas** [comment]
@fiigii Would you mind if @kasper3 or somebody else picks this up?

The hardware intrinsics project will keep you occupied for a while...

**Comment by fiigii** [comment]
@jkotas It would be great if anyone can pick this up. Sorry for the delay on this work.

**Pull request closed without merging**